zapcore/sampler.go

```diff
@@ -62,33 +69,85 @@ func fnv32a(s string) uint32 {
 	return hash
 }
 
-func (c *counter) IncCheckReset(t time.Time, tick time.Duration) uint64 {
+func (c *counter) IncCheckReset(t time.Time, tick time.Duration) (currentVal, preResetVal uint64) {
 	tn := t.UnixNano()
 	resetAfter := c.resetAt.Load()
 	if resetAfter > tn {
-		return c.counter.Inc()
+		return c.counter.Inc(), 0
 	}
 
-	c.counter.Store(1)
-
 	newResetAfter := tn + tick.Nanoseconds()
 	if !c.resetAt.CAS(resetAfter, newResetAfter) {
-		// We raced with another goroutine trying to reset, and it also reset
-		// the counter to 1, so we need to reincrement the counter.
-		return c.counter.Inc()
+		// We raced with another goroutine trying to reset, and it reset the
+		// counter to 1, so we just increment.
+		return c.counter.Inc(), 0
 	}
+	preResetVal = c.counter.Load()
+	currentVal = 1
+	c.counter.Store(currentVal)
+	return
 
-	return 1
 }
 
 type sampler struct {
 	Core
 
+	*reporting
 	counts            *counters
 	tick              time.Duration
 	first, thereafter uint64
 }
 
+type reporting struct {
+	enabled    bool
+	core       Core
+	loggerName string
+	level      Level
+	message    string
+}
+
+// NewReportingSampler creates a Core that samples incoming entries, which caps
+// the CPU and I/O load of logging while attempting to preserve a representative
+// subset of your logs. The difference with the regular sampling core is that it
+// reports the number of sampled entries for the same logging level and message.
+// Reporting is in form of a separate log entry per every tick when there are
+// sampled entries. reportingLevel must be enabled on the core.
+func NewReportingSampler(
+	core Core,
+	tick time.Duration,
+	first, thereafter int,
+	reportingLevel Level,
+	reportingLoggerName, reportingMessage string,
```

Comment:
we shouldn't take these as function args, as any additional fields we may want now require a breaking API change.

I think the implementation should be more generic rather than trying to describe every single field. E.g., maybe a function that returns a log message, or a function that can add a specific log message to the core.